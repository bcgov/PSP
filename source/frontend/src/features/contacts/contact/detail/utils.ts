import { ContactInfoField } from 'features/contacts/interfaces';
import { Dictionary } from 'interfaces/Dictionary';
import { IContactAddress, IContactMethod } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export function getContactInfo(
  validTypes: Dictionary<string>,
  contactMethods?: IContactMethod[],
): ContactInfoField[] {
  if (contactMethods === undefined) {
    return [];
  }
  // Get only the valid types
  let filteredFields = contactMethods.reduce(
    (accumulator: ContactInfoField[], method: IContactMethod) => {
      if (Object.keys(validTypes).includes(method.contactMethodType.id)) {
        accumulator.push({
          info: method.value,
          label: validTypes[method.contactMethodType.id],
        });
      }
      return accumulator;
    },
    [],
  );

  // Sort according to the dictionary order
  return filteredFields.sort((a, b) => {
    return Object.values(validTypes).indexOf(a.label) - Object.values(validTypes).indexOf(b.label);
  });
}

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
