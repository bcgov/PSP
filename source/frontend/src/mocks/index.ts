export * from './dataMocks';
export * from './filterDataMock';
export * from './mockDistrictLayerResponse';
export * from './mockFAParcelLayerResponse';
export * from './mockFinancialCode';
export * from './mockGeocoderResponses';
export * from './mockLookups';
export * from './mockMotiRegionLayerResponse';
