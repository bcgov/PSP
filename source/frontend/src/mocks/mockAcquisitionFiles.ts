import { Api_AcquisitionFile, Api_AcquisitionFileOwner } from 'models/api/AcquisitionFile';

export const mockAcquisitionFileResponse = (
  id = 1,
  name = 'Test ACQ File',
  rowVersion = 1,
): Api_AcquisitionFile => ({
  id,
  rowVersion,
  fileNumber: '1-12345-01',
  fileName: name,
  project: {
    code: '001',
    description: `Hwy 14 Expansion - Vancouver Island but it's really long so it can wrap around if it has to`,
  },
  legacyFileNumber: 'legacy file number',
  assignedDate: '2022-06-27T00:00:00',
  deliveryDate: '2022-07-29T00:00:00',
  fileStatusTypeCode: {
    id: 'ACTIVE',
    description: 'Active',
    isDisabled: false,
  },
  acquisitionTypeCode: {
    id: 'CONSEN',
    description: 'Consensual Agreement',
    isDisabled: false,
  },
  regionCode: {
    id: 1,
    description: 'South Coast Region',
    isDisabled: false,
  },
  fileProperties: [
    {
      id: 1,
      isDisabled: false,
      property: {
        id: 442,
        anomalies: [],
        tenures: [],
        roadTypes: [],
        adjacentLands: [],
        region: {
          id: 1,
          description: 'South Coast Region',
          isDisabled: false,
        },
        district: {
          id: 2,
          description: 'Vancouver Island District',
          isDisabled: false,
        },
        dataSourceEffectiveDate: '2022-08-30T00:00:00',
        latitude: 392268.0638614455,
        longitude: 1193862.2178506008,
        isSensitive: false,
        isRwyBeltDomPatent: false,
        pid: 23214937,
        landArea: 0,
        isVolumetricParcel: false,
        location: {
          coordinate: {
            x: 1193862.2178506008,
            y: 392268.0638614455,
          },
        },
        rowVersion: 1,
      },
      rowVersion: 1,
    },
    {
      id: 2,
      isDisabled: false,
      property: {
        id: 443,
        anomalies: [],
        tenures: [],
        roadTypes: [],
        adjacentLands: [],
        region: {
          id: 1,
          description: 'South Coast Region',
          isDisabled: false,
        },
        district: {
          id: 2,
          description: 'Vancouver Island District',
          isDisabled: false,
        },
        dataSourceEffectiveDate: '2022-08-30T00:00:00',
        latitude: 392208.44089083467,
        longitude: 1193818.724719974,
        isSensitive: false,
        isRwyBeltDomPatent: false,
        pid: 24996777,
        landArea: 0,
        isVolumetricParcel: false,
        location: {
          coordinate: {
            x: 1193818.724719974,
            y: 392208.44089083467,
          },
        },
        rowVersion: 1,
      },
      rowVersion: 1,
    },
  ],
  acquisitionTeam: [
    {
      id: 1,
      person: {
        id: 1,
        isDisabled: false,
        surname: 'Smith',
        firstName: 'Bob',
        middleNames: 'Billy',
        preferredName: 'Tester McTest',
        personOrganizations: [],
        personAddresses: [],
        contactMethods: [],
        comment: 'This is a test comment.',
        rowVersion: 2,
      },
      personProfileType: {
        id: 'NEGOTAGENT',
        description: 'Negotiation agent',
        isDisabled: false,
      },
      isDisabled: false,
      rowVersion: 2,
    },
    {
      id: 2,
      person: {
        id: 3,
        isDisabled: false,
        surname: 'Cheese',
        firstName: 'Stinky',
        middleNames: '',
        personOrganizations: [],
        personAddresses: [],
        contactMethods: [],
        rowVersion: 1,
      },
      personProfileType: {
        id: 'NEGOTAGENT',
        description: 'Negotiation agent',
        isDisabled: false,
      },
      isDisabled: false,
      rowVersion: 2,
    },
  ],
  acquisitionFileOwners: [
    {
      id: 2,
      acquisitionFileId: 2,
      address: {
        id: 4,
        streetAddress1: '456 Souris Street',
        streetAddress2: 'PO Box 250',
        streetAddress3: 'A Hoot and a holler from the A&W',
        municipality: 'North Podunk',
        provinceStateId: 1,
        countryId: 1,
        postal: 'IH8 B0B',
        rowVersion: 1,
      },
      lastNameOrCorp1: 'john',
      lastNameOrCorp2: 'Doe',
      givenName: 'Sr.',
      appCreateTimestamp: '2023-02-06T22:27:22.01',
      appLastUpdateTimestamp: '2023-02-06T22:27:22.01',
      appLastUpdateUserid: 'dbo',
      appCreateUserid: 'dbo',
      rowVersion: 3,
    },
    {
      id: 3,
      acquisitionFileId: 2,
      address: {
        id: 3,
        streetAddress1: '123 Main Street',
        streetAddress2: 'PO Box 123',
        streetAddress3: 'Next to the Dairy Queen',
        municipality: 'East Podunk',
        provinceStateId: 1,
        countryId: 1,
        postal: 'I4M B0B',
        rowVersion: 1,
      },
      lastNameOrCorp1: 'FORTIS BC',
      lastNameOrCorp2: 'LTD',
      incorporationNumber: '9999',
      appCreateTimestamp: '2023-02-08T18:11:12.303',
      appLastUpdateTimestamp: '2023-02-08T18:11:12.303',
      appLastUpdateUserid: 'dbo',
      appCreateUserid: 'dbo',
      rowVersion: 2,
    },
  ],
  appCreateTimestamp: '2022-05-28T00:57:37.42',
  appLastUpdateTimestamp: '2022-07-28T00:57:37.42',
  appLastUpdateUserid: 'admin',
  appCreateUserid: 'admin',
  appLastUpdateUserGuid: '14c9a273-6f4a-4859-8d59-9264d3cee53f',
  appCreateUserGuid: '14c9a273-6f4a-4859-8d59-9264d3cee53f',
});

export const mockAcquisitionFileOwnersResponse = (
  acquistionFileId: number = 1,
): Api_AcquisitionFileOwner[] => [
  {
    id: 1,
    acquisitionFileId: acquistionFileId,
    address: {
      id: 4,
      streetAddress1: '456 Souris Street',
      streetAddress2: 'PO Box 250',
      streetAddress3: 'A Hoot and a holler from the A\u0026W',
      municipality: 'North Podunk',
      provinceStateId: 1,
      province: { id: 1, code: 'BC', description: 'British Columbia', displayOrder: 10 },
      countryId: 1,
      country: { id: 1, code: 'CA', description: 'Canada', displayOrder: 1 },
      postal: 'IH8 B0B',
      rowVersion: 1,
    },
    lastNameOrCorp1: 'JOHH',
    lastNameOrCorp2: 'DOE',
    givenName: 'Sr.',
    appCreateTimestamp: '0001-01-01T00:00:00',
    appLastUpdateTimestamp: '0001-01-01T00:00:00',
    rowVersion: 3,
  },
  {
    id: 3,
    acquisitionFileId: acquistionFileId,
    address: {
      id: 3,
      streetAddress1: '123 Main Street',
      streetAddress2: 'PO Box 123',
      streetAddress3: 'Next to the Dairy Queen',
      municipality: 'East Podunk',
      provinceStateId: 1,
      province: { id: 1, code: 'BC', description: 'British Columbia', displayOrder: 10 },
      countryId: 1,
      country: { id: 1, code: 'CA', description: 'Canada', displayOrder: 1 },
      postal: 'I4M B0B',
      rowVersion: 1,
    },
    lastNameOrCorp1: 'FORTIS BC',
    lastNameOrCorp2: 'LTD',
    incorporationNumber: '9999',
    appCreateTimestamp: '0001-01-01T00:00:00',
    appLastUpdateTimestamp: '0001-01-01T00:00:00',
    rowVersion: 2,
  },
];
