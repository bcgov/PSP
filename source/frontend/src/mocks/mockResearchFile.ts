import { Api_ResearchFile } from 'models/api/ResearchFile';
export const getMockResearchFile = (): Api_ResearchFile => ({
  id: 109,
  fileName: 'fsdgsdf',
  fileNumber: 'R-109',
  roadName: 'Test road name',
  roadAlias: 'Test road alias',
  fileStatusTypeCode: {
    id: 'ACTIVE',
    description: 'Active',
    isDisabled: false,
  },
  fileProperties: [
    {
      id: 55,
      isDisabled: false,
      property: {
        id: 495,
        anomalies: [],
        tenures: [],
        roadTypes: [],
        adjacentLands: [],
        region: {
          id: 1,
          description: 'South Coast Region',
          isDisabled: false,
        },
        district: {
          id: 1,
          description: 'Lower Mainland District',
          isDisabled: false,
        },
        dataSourceEffectiveDate: '2022-09-02T00:00:00',
        latitude: 49.27720127104871,
        longitude: -123.128633565,
        isSensitive: false,
        isRwyBeltDomPatent: false,
        pin: 0,
        pid: 123456789,
        planNumber: 'EPP92028',
        landArea: 0,
        isVolumetricParcel: false,
        location: {
          coordinate: {
            x: -123.128633565,
            y: 49.27720127104871,
          },
        },
        rowVersion: 1,
      },
      rowVersion: 1,
    },
  ],
  requestDate: '2022-04-14T00:00:00',
  requestDescription: 'a request description',
  researchResult: 'A research result',
  researchCompletionDate: '2022-03-30T00:00:00',
  isExpropriation: false,
  expropriationNotes: 'An expropriation note',
  requestSourceType: {
    id: 'HQ',
    description: 'Headquarters (HQ)',
    isDisabled: false,
  },
  requestorOrganization: {
    id: 3,
    isDisabled: false,
    name: 'Dairy Queen Forever! Property Management',
    organizationPersons: [],
    organizationAddresses: [],
    contactMethods: [],
    rowVersion: 1,
  },
  researchFilePurposes: [],
  researchFileProjects: [
    {
      id: 1,
      isDisabled: false,
      project: {
        id: 1,
        code: '00048',
        description: 'CLAIMS',
        products: [],
        appCreateTimestamp: '2023-01-30T21:33:33.063',
        appLastUpdateTimestamp: '2023-01-30T21:33:33.063',
        appLastUpdateUserid: 'dbo',
        appCreateUserid: 'dbo',
        rowVersion: 1,
      },
      fileId: 109,
      rowVersion: 1,
    },
  ],
  appCreateTimestamp: '2022-09-02T19:29:50.397',
  appLastUpdateTimestamp: '2022-09-02T19:29:50.397',
  appLastUpdateUserid: 'admin',
  appCreateUserid: 'admin',
  appLastUpdateUserGuid: '14c9a273-6f4a-4859-8d59-9264d3cee53f',
  appCreateUserGuid: '14c9a273-6f4a-4859-8d59-9264d3cee53f',
  rowVersion: 1,
});
