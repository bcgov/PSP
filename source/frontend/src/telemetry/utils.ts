import { metrics, trace } from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { browserDetector } from '@opentelemetry/opentelemetry-browser-detector';
import { detectResourcesSync, Resource, ResourceAttributes } from '@opentelemetry/resources';
import { MeterProvider, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import {
  BatchSpanProcessor,
  ConsoleSpanExporter,
  SimpleSpanProcessor,
  SpanProcessor,
  WebTracerProvider,
} from '@opentelemetry/sdk-trace-web';
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
  ATTR_USER_AGENT_ORIGINAL,
} from '@opentelemetry/semantic-conventions';
import {
  ATTR_DEPLOYMENT_ENVIRONMENT_NAME,
  ATTR_SERVICE_INSTANCE_ID,
} from '@opentelemetry/semantic-conventions/incubating';
import isAbsoluteUrl from 'is-absolute-url';
import { v4 as uuidv4 } from 'uuid';

import { TelemetryConfig } from './config';
import { BrowserAttributesSpanProcessor } from './traces/BrowserAttributesSpanProcessor';
import { UserInfoSpanProcessor } from './traces/UserInfoSpanProcessor';

export const isBrowserEnvironment = () => {
  return typeof window !== 'undefined';
};

export const isNodeEnvironment = () => {
  return typeof process !== 'undefined' && process.release && process.release.name === 'node';
};

// creates URL and appends query parameters
export const buildUrl = (inputUrl: string, queryParams: Record<string, any> = {}): URL => {
  const baseUrl = window.location.origin;
  const urlInstance = isAbsoluteUrl(inputUrl) ? new URL(inputUrl) : new URL(inputUrl, baseUrl);
  Object.keys(queryParams).forEach(k => {
    if (queryParams[k] !== undefined) {
      urlInstance.searchParams.set(k, queryParams[k]);
    }
  });
  return urlInstance;
};

export const isBlocked = (uri: string, config: TelemetryConfig) => {
  const blockList = [...(config.denyUrls ?? []), config.otlpEndpoint];
  return blockList.findIndex(blocked => uri.includes(blocked)) >= 0;
};

// List of meters in the application: e.g. "network", "webvitals", "app", etc
export const NETWORK_METER = 'network-meter';

const makeResource = (config: TelemetryConfig, extraAttributes?: ResourceAttributes) => {
  const uuid = uuidv4();
  let resource = new Resource({
    [ATTR_DEPLOYMENT_ENVIRONMENT_NAME]: config?.environment,
    [ATTR_SERVICE_NAME]: config?.name,
    [ATTR_SERVICE_VERSION]: config?.appVersion,
    [ATTR_SERVICE_INSTANCE_ID]: uuid,
    [ATTR_USER_AGENT_ORIGINAL]: typeof navigator !== 'undefined' ? navigator.userAgent : '',
    'session.instance.id': uuid,
    'screen.width': window.screen.width,
    'screen.height': window.screen.height,
    ...extraAttributes,
  });

  const detectedResources = detectResourcesSync({ detectors: [browserDetector] });
  resource = resource.merge(detectedResources);
  return resource;
};

export const registerMeterProvider = (
  config: TelemetryConfig,
  extraAttributes?: ResourceAttributes,
) => {
  if (config.debug) {
    console.info('[INFO] Registering metrics provider');
  }

  // This is common metadata sent with every metric measurement
  const resource = makeResource(config, extraAttributes);
  const metricExporter = new OTLPMetricExporter({
    url: new URL('/v1/metrics', config.otlpEndpoint).href,
  });

  const meterProvider = new MeterProvider({
    resource: resource,
    readers: [
      new PeriodicExportingMetricReader({
        exporter: metricExporter,
        exportIntervalMillis: config?.exportInterval || 30_000, // export metrics every 30s by default
      }),
    ],
  });

  // set this MeterProvider to be global to the app being instrumented.
  metrics.setGlobalMeterProvider(meterProvider);
};

export const registerTracerProvider = (
  config: TelemetryConfig,
  extraAttributes?: ResourceAttributes,
) => {
  if (config.debug) {
    console.info('[INFO] Registering trace provider');
  }

  // This is common metadata sent with every trace
  const resource = makeResource(config, extraAttributes);
  const exporter = new OTLPTraceExporter({
    url: new URL('v1/traces', config.otlpEndpoint).href,
  });

  const processors: SpanProcessor[] = [];

  if (config.debug) {
    processors.push(new SimpleSpanProcessor(new ConsoleSpanExporter()));
  }

  // use the batch processor for better performance
  processors.push(
    new BatchSpanProcessor(exporter, { scheduledDelayMillis: config?.exportInterval || 5000 }), // export traces every 5s by default
    new BrowserAttributesSpanProcessor(),
    new UserInfoSpanProcessor(),
  );

  const provider = new WebTracerProvider({
    resource,
    spanProcessors: [...processors],
  });

  // set up context propagation
  provider.register({
    propagator: new W3CTraceContextPropagator(),
  });

  // set this TracerProvider to be global to the app
  trace.setGlobalTracerProvider(provider);
};
