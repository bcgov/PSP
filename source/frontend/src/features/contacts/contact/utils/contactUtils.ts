import { CountryCodes } from 'constants/index';
import { AddressField } from 'features/contacts/interfaces';
import { validateYupSchema, yupToFormErrors } from 'formik';
import { IEditableOrganizationForm, IEditablePersonForm } from 'interfaces/editable-contact';
import isEmpty from 'lodash/isEmpty';

import {
  hasAddress,
  hasEmail,
  hasPhoneNumber,
  OrganizationValidationSchema,
  PersonValidationSchema,
} from '../create/validation';
import { sortAddresses } from '../detail/utils';
import { IContactAddress } from './../../../../interfaces/IContact';

export const onValidateOrganization = (
  values: IEditableOrganizationForm,
  otherCountryId?: string,
) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, OrganizationValidationSchema, true, {
      otherCountry: otherCountryId,
    });
    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const onValidatePerson = (values: IEditablePersonForm, otherCountryId?: string) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, PersonValidationSchema, true, { otherCountry: otherCountryId });

    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const toAddressFields = (addresses: IContactAddress[]) => {
  return addresses
    .sort(sortAddresses)
    .reduce((accumulator: AddressField[], value: IContactAddress) => {
      accumulator.push({
        label: value.addressType.description || '',
        streetAddress1: value.streetAddress1,
        streetAddress2: value.streetAddress2,
        streetAddress3: value.streetAddress3,
        municipalityAndProvince: [
          value.municipality !== undefined ? value.municipality : '',
          value.province?.provinceStateCode ?? '',
        ]
          .filter(x => !isEmpty(x))
          .join(' '),
        country:
          value.country?.countryCode === CountryCodes.Other
            ? value.countryOther ?? ''
            : value.country?.description,
        postal: value.postal,
      });
      return accumulator;
    }, []);
};
