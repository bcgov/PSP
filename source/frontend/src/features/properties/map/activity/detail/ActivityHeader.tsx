import { StyledDivider } from 'components/common/styles';
import { UserNameTooltip } from 'components/common/UserNameTooltip';
import { HeaderField } from 'features/mapSideBar/tabs/HeaderField';
import { Api_Activity } from 'models/api/Activity';
import { Api_File } from 'models/api/File';
import * as React from 'react';
import { Col, Row } from 'react-bootstrap';
import { prettyFormatDate } from 'utils';

export interface IActivityHeaderProps {
  file: Api_File;
  activity: Api_Activity;
}

export const ActivityHeader: React.FunctionComponent<IActivityHeaderProps> = ({
  file,
  activity,
}) => {
  const leftColumnWidth = '5';
  const leftColumnLabel = '3';
  return (
    <Row className="no-gutters">
      <Col>
        <Row className="no-gutters">
          <Col xs={leftColumnWidth}>
            <HeaderField label="File:" labelWidth={leftColumnLabel} className="flex-nowrap">
              {file?.fileNumber} - {file?.fileName}
            </HeaderField>
          </Col>
          <Col className="text-right">
            Created: <strong>{prettyFormatDate(activity?.appCreateTimestamp)}</strong> by{' '}
            <UserNameTooltip
              userName={activity?.appCreateUserid}
              userGuid={activity?.appCreateUserGuid}
            />
          </Col>
        </Row>
        <Row className="no-gutters">
          <Col xs={leftColumnWidth}></Col>
          <Col className="text-right">
            Last updated: <strong>{prettyFormatDate(activity?.appLastUpdateTimestamp)}</strong> by{' '}
            <UserNameTooltip
              userName={activity?.appLastUpdateUserid}
              userGuid={activity?.appLastUpdateUserGuid}
            />
          </Col>
        </Row>
        <Row className="no-gutters">
          <Col xs={leftColumnWidth}></Col>
          <Col>
            <HeaderField className="justify-content-end" label="Status:">
              {activity?.activityStatusTypeCode?.description}
            </HeaderField>
          </Col>
        </Row>
        <StyledDivider />
      </Col>
    </Row>
  );
};

export default ActivityHeader;
