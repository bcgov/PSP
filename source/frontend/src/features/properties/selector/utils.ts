import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { formatApiAddress, pidFormatter } from 'utils';

import { Api_Geometry, Api_Property } from './../../../models/api/Property';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
  ADDRESS = 'Address',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.longitude?.toFixed(6), property.latitude?.toFixed(6)]).join(', '),
    };
  } else if (property.address !== undefined) {
    return {
      label: NameSourceType.ADDRESS,
      value: property.address,
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getPrettyLatLng = (location?: Api_Geometry) =>
  compact([
    location?.coordinate?.x?.toFixed(6) ?? 0,
    location?.coordinate?.y?.toFixed(6) ?? 0,
  ]).join(', ');

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    return getApiPropertyName(property);
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getApiPropertyName = (property?: Api_Property): PropertyName => {
  if (property === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  let mapProperty: IMapProperty = {
    pin: property.pin?.toString(),
    pid: property.pid?.toString(),
    latitude: property.latitude,
    longitude: property.longitude,
    planNumber: property.planNumber,
    address: property.address !== undefined ? formatApiAddress(property.address) : undefined,
  };
  return getPropertyName(mapProperty);
};
