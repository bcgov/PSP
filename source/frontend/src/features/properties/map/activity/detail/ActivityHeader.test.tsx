import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { mockAcquisitionFileResponse } from 'mocks/mockAcquisitionFiles';
import { getMockActivityResponse } from 'mocks/mockActivities';
import { getMockResearchFile } from 'mocks/mockResearchFile';
import { prettyFormatDate } from 'utils';
import { render, RenderOptions } from 'utils/test-utils';

import ActivityHeader, { IActivityHeaderProps } from './ActivityHeader';

const mockAxios = new MockAdapter(axios);

describe('ActivityHeader component', () => {
  // render component under test
  const setup = (props: Partial<IActivityHeaderProps>, renderOptions: RenderOptions = {}) => {
    const utils = render(
      <ActivityHeader
        file={props.file ?? { ...getMockResearchFile(), id: 1 }}
        activity={props?.activity ?? { ...getMockActivityResponse(), id: 2 }}
      />,
      {
        ...renderOptions,
      },
    );

    return { ...utils };
  };

  beforeEach(() => {
    mockAxios.onGet(new RegExp('users/info/*')).reply(200, {});
  });

  afterEach(() => {
    mockAxios.reset();
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    const { asFragment } = setup({ file: getMockResearchFile() });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when an acquisition file is provided', async () => {
    const testAcquisitionFile = { ...mockAcquisitionFileResponse(), id: 1 };
    const testActivity = getMockActivityResponse();
    const { getByText } = setup({
      file: { ...mockAcquisitionFileResponse(), id: 1 },
      activity: testActivity,
    });

    expect(
      getByText(`${testAcquisitionFile.fileNumber as string} - ${testAcquisitionFile.fileName}`),
    ).toBeVisible();
    expect(getByText(prettyFormatDate(testActivity.appCreateTimestamp))).toBeVisible();
    expect(getByText(prettyFormatDate(testActivity.appLastUpdateTimestamp))).toBeVisible();
    expect(getByText(testActivity?.activityStatusTypeCode?.description ?? '')).toBeVisible();
  });

  it('renders as expected when a research file is provided', async () => {
    const testResearchFile = getMockResearchFile();
    const testActivity = getMockActivityResponse();
    const { getByText } = setup({ file: testResearchFile, activity: testActivity });

    expect(
      getByText(`${testResearchFile.fileNumber as string} - ${testResearchFile.fileName}`),
    ).toBeVisible();
    expect(getByText(prettyFormatDate(testActivity.appCreateTimestamp))).toBeVisible();
    expect(getByText(prettyFormatDate(testActivity.appLastUpdateTimestamp))).toBeVisible();
    expect(getByText(testActivity?.activityStatusTypeCode?.description ?? '')).toBeVisible();
  });
});
