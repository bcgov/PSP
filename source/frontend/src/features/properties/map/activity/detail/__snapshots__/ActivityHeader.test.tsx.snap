// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityHeader component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container"
  />
  <div
    class="Toastify"
  />
  .c1 {
  margin-top: 0.5rem;
  margin-bottom: 1.5rem;
  border-bottom-style: solid;
  border-bottom-color: grey;
  border-bottom-width: 0.1rem;
}

.c0 {
  font-weight: bold;
}

<div
    class="no-gutters row"
  >
    <div
      class="col"
    >
      <div
        class="no-gutters row"
      >
        <div
          class="col-5"
        >
          <div
            class="flex-nowrap row"
          >
            <div
              class="pr-0 text-left col-3"
            >
              <label>
                File:
              </label>
            </div>
            <div
              class="pl-1 text-left col-auto"
            >
              <strong>
                R-109 - fsdgsdf
              </strong>
            </div>
          </div>
        </div>
        <div
          class="text-right col"
        >
          Created: 
          <strong>
            Jan 1, 0001
          </strong>
           by 
          <span
            class="tooltip-icon"
            data-testid="tooltip-icon"
            id="userNameTooltip"
          >
            <span
              class="c0"
            >
              USER
            </span>
          </span>
        </div>
      </div>
      <div
        class="no-gutters row"
      >
        <div
          class="col-5"
        />
        <div
          class="text-right col"
        >
          Last updated: 
          <strong>
            Feb 1, 0001
          </strong>
           by 
          <span
            class="tooltip-icon"
            data-testid="tooltip-icon"
            id="userNameTooltip"
          >
            <span
              class="c0"
            >
              USER
            </span>
          </span>
        </div>
      </div>
      <div
        class="no-gutters row"
      >
        <div
          class="col-5"
        />
        <div
          class="col"
        >
          <div
            class="justify-content-end row"
          >
            <div
              class="pr-0 text-left col-auto"
            >
              <label>
                Status:
              </label>
            </div>
            <div
              class="pl-1 text-left col-auto"
            >
              <strong>
                Not Started
              </strong>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c1"
      />
    </div>
  </div>
</DocumentFragment>
`;
